### ISSUE TITLE
defined-ness of substrings disappear over repeated calls

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8469]
Migrated from [rt.perl.org#39247](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39247) (status was 'resolved')

Searchable as RT39247$

**Comment by p5pRT** [comment]
### From martin@datacash.com

#### Created by martin@datacash.com



I have experienced a problem with data values disappearing when using external 
modules like DBD&#8203;::mysql in certain situations\.

I have narrowed this down to a problem with the FLAGS for SVs derived 
from substrings within subroutines\.  This can be seen with a simple 
script using Devel&#8203;::Peek&#8203;:

=======================================================
\#\! /usr/bin/perl5 \-w

use strict;
use Devel&#8203;::Peek;

my $v = '4000930012345671';

print "=" x 20\, "Call 1"\, "=" x 20\, "\\n";
cancel\( $v \);  \# pPOK flag is set

print "=" x 20\, "Call 2"\, "=" x 20\, "\\n";
cancel\( $v \);  \# bug \- pPOK flag is not set

sub cancel \{
&nbsp;  my \( $dcref \) = @&#8203;\_;
&nbsp;  Dump\( substr\( $dcref\, \-11 \) \);
\}
======================================================


The significant lines of output are \(on my machine\)&#8203;:
====================Call 1====================
SV = PVLV\(0x81634e8\) at 0x81627f0
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(PADMY\,GMG\,SMG\,pPOK\)
\.\.\.
====================Call 2====================
SV = PVLV\(0x81634e8\) at 0x81627f0
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(PADMY\,GMG\,SMG\)

The problem is that there is no pPOK \(or any other \*OK\) 
in the FLAGS for the second call to cancel\(\)\.  

This causes a particular problem for me with the DBD&#8203;::mysql module\,
where values supplied to the do method are tested using SvOK\(\) to determine
whether the value should be extracted from the SV or whether they should
be set to NULL\.   In my application\, this occasionally causes values to 
appear as NULL where non\-null values have been supplied\.

I have experienced this problem with Perl 5\.8\.4\, and I have been told
that it is also a problem with Perl 5\.8\.3\.  I have compiled the latest
version 5\.8\.8\, and the problem appears to be there also\.

In Perl 5\.6\.1 \(which is my current production environment\) the 
same script produces&#8203;:

====================Call 1====================
SV = PVLV\(0x80ff470\) at 0x8115d64
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(PADMY\,GMG\,SMG\,pPOK\)
\.\.\.
====================Call 2====================
SV = PVLV\(0x80ff470\) at 0x8115d64
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(PADMY\,GMG\,SMG\,POK\,pPOK\)

Again\, the FLAGS are different between calls \(I do not understand why\)\,
but at least the pPOK flag remains set\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.8:

Configured by martin at Tue May 30 18:37:36 BST 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.4.31-ibm-p4-xeon-lsi-2, archname=i686-linux
    uname='linux chadwick 2.4.31-ibm-p4-xeon-lsi-2 #1 smp fri aug 26 16:51:28 bst 2005 i686 gnulinux '
    config_args='-Dcc=gcc -d'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='3.3.5 (Debian 1:3.3.5-13)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /usr/local/lib/perl5/5.8.8/i686-linux
    /usr/local/lib/perl5/5.8.8
    /usr/local/lib/perl5/site_perl/5.8.8/i686-linux
    /usr/local/lib/perl5/site_perl/5.8.8
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.8:
    HOME=/home/martin
    LANG=en_GB
    LANGUAGE=en_GB:en_US:en_GB:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From BQW10602@nifty.com


> \# New Ticket Created by  martin@&#8203;datacash\.com 
> \# Please include the string&#8203;:  \[perl \#39247\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39247 >

> The significant lines of output are \(on my machine\)&#8203;:
> ====================Call 1====================
> SV = PVLV\(0x81634e8\) at 0x81627f0
>   REFCNT = 1
>   FLAGS = \(PADMY\,GMG\,SMG\,pPOK\)
> \.\.\.
> ====================Call 2====================
> SV = PVLV\(0x81634e8\) at 0x81627f0
>   REFCNT = 1
>   FLAGS = \(PADMY\,GMG\,SMG\)
> 
> The problem is that there is no pPOK \(or any other \*OK\) 
> in the FLAGS for the second call to cancel\(\)\.  

What turns pPOK off in the case of Call 2 is SvOK\_off below\,
that is introduced at Change \#20462 \(\[PATCH\] Re&#8203;: \[perl \#23207\] persistant
sideffect involving bitwise xor and substr\)\.

The reversion of Change \#20462 as below makes it POK\,pPOK\.
It's a wonder that the reversion doesn't cause the problem
in perl \#23207 to come back\.

<details><summary>Inline Patch</summary>

```diff
--- perl~/pp.c	Mon May 15 17:38:36 2006
+++ perl/pp.c	Wed May 31 21:37:28 2006
@@ -3126,8 +3126,6 @@
 		sv_upgrade(TARG, SVt_PVLV);
 		sv_magic(TARG, NULL, PERL_MAGIC_substr, NULL, 0);
 	    }
-	    else
-		SvOK_off(TARG);
 
 	    LvTYPE(TARG) = 'x';
 	    if (LvTARG(TARG) != sv) {


```
</details>
> In Perl 5.6.1 (which is my current production environment) the 
> same script produces&#8203;:
> 
> ====================Call 1====================
> SV = PVLV\(0x80ff470\) at 0x8115d64
>   REFCNT = 1
>   FLAGS = \(PADMY\,GMG\,SMG\,pPOK\)
> \.\.\.
> ====================Call 2====================
> SV = PVLV\(0x80ff470\) at 0x8115d64
>   REFCNT = 1
>   FLAGS = \(PADMY\,GMG\,SMG\,POK\,pPOK\)
> 
> Again\, the FLAGS are different between calls \(I do not understand why\)\,

The sv\_setpvn\(TARG\, tmps\, rem\) in Perl\_pp\_substr\(\) turns POK of TARG on\.
But replacement with sv\_setpvn\_mg seems to break TARG\,
as commented at the end of Perl\_pp\_substrr\(\)&#8203;:

&nbsp;    PUSHs\(TARG\);        /\* avoid SvSETMAGIC here \*/

> but at least the pPOK flag remains set\.

Here is a possible test suite for pPOK remaining\. But perhaps this test
is incorrect as it assumes that sv\_utf8\_decode\(\) neglects SvGETMAGIC\(\)
before SvPOKp\(\)\.

\#\!perl
use Test;
plan tests => 2;

sub cancel \{ ok\(\!utf8&#8203;::decode\(substr\($\_\[0\]\, \-2\)\)\) \}

my $v = "abc\\x\{100\}";
cancel\($v\); 
cancel\($v\); \# perl\-current fails here
\_\_END\_\_

Regards\,
SADAHIRO Tomoyuki

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From martin@datacash.com

SADAHIRO Tomoyuki via RT wrote&#8203;:

>>\# New Ticket Created by  martin@&#8203;datacash\.com 
>>\# Please include the string&#8203;:  \[perl \#39247\]
>>\# in the subject line of all future correspondence about this issue\. 
>>\# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39247 >
>>    
>>
>
>  
>
>>The significant lines of output are \(on my machine\)&#8203;:
>>====================Call 1====================
>>SV = PVLV\(0x81634e8\) at 0x81627f0
>>  REFCNT = 1
>>  FLAGS = \(PADMY\,GMG\,SMG\,pPOK\)
>>\.\.\.
>>====================Call 2====================
>>SV = PVLV\(0x81634e8\) at 0x81627f0
>>  REFCNT = 1
>>  FLAGS = \(PADMY\,GMG\,SMG\)
>>
>>The problem is that there is no pPOK \(or any other \*OK\) 
>>in the FLAGS for the second call to cancel\(\)\.  
>>    
>>
>
>What turns pPOK off in the case of Call 2 is SvOK\_off below\,
>that is introduced at Change \#20462 \(\[PATCH\] Re&#8203;: \[perl \#23207\] persistant
>sideffect involving bitwise xor and substr\)\.
>
>The reversion of Change \#20462 as below makes it POK\,pPOK\.
>It's a wonder that the reversion doesn't cause the problem
>in perl \#23207 to come back\.
>
>  
>
In stable \(5\.8\.8\)\, the problem in perl \#23207 does come back if I revert
Change \#20462 :

t/op/substr\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\# Failed at op/substr\.t line 613
\#      got 'BA'
\# expected 'A'
FAILED at test 318

Which version of Perl did you apply the reversion to ?


regards\,
Martin

**Comment by p5pRT** [comment]
### From BQW10602@nifty.com


On Wed\, 31 May 2006 17&#8203;:57&#8203;:32 \+0100\, Martin Waite \<martin@&#8203;datacash\.com> wrote

> SADAHIRO Tomoyuki via RT wrote&#8203;:
> 
> >>\# New Ticket Created by  martin@&#8203;datacash\.com 
> >>\# Please include the string&#8203;:  \[perl \#39247\]
> >>\# in the subject line of all future correspondence about this issue\. 
> >>\# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39247 >

> >The reversion of Change \#20462 as below makes it POK\,pPOK\.
> >It's a wonder that the reversion doesn't cause the problem
> >in perl \#23207 to come back\.
> >
> >  
> >
> In stable \(5\.8\.8\)\, the problem in perl \#23207 does come back if I revert
> Change \#20462 :
> 
> t/op/substr\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\# Failed at op/substr\.t line 613
> \#      got 'BA'
> \# expected 'A'
> FAILED at test 318
> 
> Which version of Perl did you apply the reversion to ?

OK     when the reversion is applied to 5\.9\.\[1\-3\]\.
Not OK when the reversion is applied to 5\.9\.0\.

Regards\,
SADAHIRO Tomoyuki

**Comment by p5pRT** [comment]
### From BQW10602@nifty.com


On Thu\, 01 Jun 2006 08&#8203;:16&#8203;:51 \+0900\, SADAHIRO Tomoyuki \<bqw10602@&#8203;nifty\.com> wrote

> On Wed\, 31 May 2006 17&#8203;:57&#8203;:32 \+0100\, Martin Waite \<martin@&#8203;datacash\.com> wrote
> 
> > SADAHIRO Tomoyuki via RT wrote&#8203;:
> > 
> > >>\# New Ticket Created by  martin@&#8203;datacash\.com 
> > >>\# Please include the string&#8203;:  \[perl \#39247\]
> > >>\# in the subject line of all future correspondence about this issue\. 
> > >>\# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39247 >
> 
> > >The reversion of Change \#20462 as below makes it POK\,pPOK\.
> > >It's a wonder that the reversion doesn't cause the problem
> > >in perl \#23207 to come back\.
> > 
> > In stable \(5\.8\.8\)\, the problem in perl \#23207 does come back if I revert
> > Change \#20462 :
> > 
> > t/op/substr\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\# Failed at op/substr\.t line 613
> > \#      got 'BA'
> > \# expected 'A'
> > FAILED at test 318
> > 
> > Which version of Perl did you apply the reversion to ?
> 
> OK     when the reversion is applied to 5\.9\.\[1\-3\]\.
> Not OK when the reversion is applied to 5\.9\.0\.

For bug \#23207\, both pp\_substr without "magic set" and pp\_bit\_xor
without "magic get" should be to blame\. Therefore I consider the real
fix of the bug is Change \#22074 where "magic get" is added to pp\_bit\_xor\.

&nbsp;    \[ 22074\] Some of the bitwise manipulation PP functions
&nbsp;    weren't checking whether their arguments were magical before
&nbsp;    using them\. \(bug \[\#24816\]\)\.

Though unsetting POK by Change \#20462 certainly makes a trigger to
mg\_get\(\) in sv\_2pv\_flag\(\)\, it is an overkill that POKp is also unset\.

As Change \#22074 has been integrated into the maintenance track as
Change \#27391\, perl 5\.8\.9\-to\-be can take the reversion of Change \#20462\.

Regards\,
SADAHIRO Tomoyuki

<details><summary>Inline Patch</summary>

```diff
--- perl~/pp.c	Mon May 15 17:38:36 2006
+++ perl/pp.c	Wed May 31 21:37:28 2006
@@ -3126,8 +3126,6 @@
 		sv_upgrade(TARG, SVt_PVLV);
 		sv_magic(TARG, NULL, PERL_MAGIC_substr, NULL, 0);
 	    }
-	    else
-		SvOK_off(TARG);
 
 	    LvTYPE(TARG) = 'x';
 	    if (LvTARG(TARG) != sv) {
```
</details>

**Comment by p5pRT** [comment]
### From martin@datacash.com

SADAHIRO Tomoyuki via RT wrote&#8203;:

>On Wed\, 31 May 2006 17&#8203;:57&#8203;:32 \+0100\, Martin Waite \<martin@&#8203;datacash\.com> wrote
>
>  
>
>>SADAHIRO Tomoyuki via RT wrote&#8203;:
>>
>>    
>>
>>>>\# New Ticket Created by  martin@&#8203;datacash\.com 
>>>>\# Please include the string&#8203;:  \[perl \#39247\]
>>>>\# in the subject line of all future correspondence about this issue\. 
>>>>\# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39247 >
>>>>        
>>>>
>
>  
>
>>>The reversion of Change \#20462 as below makes it POK\,pPOK\.
>>>It's a wonder that the reversion doesn't cause the problem
>>>in perl \#23207 to come back\.
>>>
>>> 
>>>
>>>      
>>>
>>In stable \(5\.8\.8\)\, the problem in perl \#23207 does come back if I revert
>>Change \#20462 :
>>
>>t/op/substr\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\# Failed at op/substr\.t line 613
>>\#      got 'BA'
>>\# expected 'A'
>>FAILED at test 318
>>
>>Which version of Perl did you apply the reversion to ?
>>    
>>
>
>OK     when the reversion is applied to 5\.9\.\[1\-3\]\.
>Not OK when the reversion is applied to 5\.9\.0\.
>
>
>  
>
Many thanks for your help\.

I wanted to ask why the defined\-ness fragility is only really visible to
the XS module I am using \(DBD&#8203;::MySQL\)\. For example Perl's defined
operator doesn't have the same problem\.  I looked in the definition of
pp\_defined and in this case the Perl defined operator uses the sequence&#8203;:

&nbsp;        if \(SvGMAGICAL\(sv\)\)
&nbsp;            mg\_get\(sv\);
&nbsp;        if \(SvOK\(sv\)\)
&nbsp;            RETPUSHYES;

However\, in the DBD&#8203;::MySQL module\, only the SvOK\(\) test is applied\.

Should the DBD&#8203;::MySQL module be modified to include the MAGIC handling
to be more robust ?

regards\,

Martin

**Comment by p5pRT** [comment]
### From BQW10602@nifty.com


On Thu\, 01 Jun 2006 11&#8203;:28&#8203;:00 \+0100\, Martin Waite \<martin@&#8203;datacash\.com>
wrote
> I wanted to ask why the defined\-ness fragility is only really visible to
> the XS module I am using \(DBD&#8203;::MySQL\)\. For example Perl's defined
> operator doesn't have the same problem\.  I looked in the definition of
> pp\_defined and in this case the Perl defined operator uses the sequence&#8203;:
> 
>         if \(SvGMAGICAL\(sv\)\)
>             mg\_get\(sv\);
>         if \(SvOK\(sv\)\)
>             RETPUSHYES;
> 
> However\, in the DBD&#8203;::MySQL module\, only the SvOK\(\) test is applied\.
> 
> Should the DBD&#8203;::MySQL module be modified to include the MAGIC handling
> to be more robust ?

Certainly both perlapi\.pod and perlguts\.pod mentions SvOK\(\) is used
to check whether an SV is defined\. But it may be necessary to call
SvGETMAGIC\(\) \(the public macro for \{if \(SvGMAGICAL\(sv\)\) mg\_get\(sv\);\}\)
before SvOK\(\)\.
For example\, mg\_get\(\) for a tied scalar corresponds to FETCH\(\) and
tie\(\) doesn't call mg\_get\(\)\. Hence a tied scalar just after tie\(\)
isn't true for SvOK\(\)\. But if FETCH\(\) has some side effect\,
calling too many mg\_get\(\) may cause some unexpected result\. Thus
I think SvGETMAGIC\(\) should not be used just before each appearance
of SvOK\(\) or other macros\.

In my opinion\, the perl's internal is too complicated for XS authors
to hack and investigate properly\. It may be a workaround to pass simple
values to XS functions always\. For example\, store the substring into
a temporary scalar variable and pass the scalar to XS\, rather than pass
substr\(\) directly to XS\.

Regards\,
SADAHIRO Tomoyuki

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgarcia

On 01/06/06\, SADAHIRO Tomoyuki \<bqw10602@&#8203;nifty\.com> wrote&#8203;:
> For bug \#23207\, both pp\_substr without "magic set" and pp\_bit\_xor
> without "magic get" should be to blame\. Therefore I consider the real
> fix of the bug is Change \#22074 where "magic get" is added to pp\_bit\_xor\.
>
>     \[ 22074\] Some of the bitwise manipulation PP functions
>     weren't checking whether their arguments were magical before
>     using them\. \(bug \[\#24816\]\)\.
>
> Though unsetting POK by Change \#20462 certainly makes a trigger to
> mg\_get\(\) in sv\_2pv\_flag\(\)\, it is an overkill that POKp is also unset\.
>
> As Change \#22074 has been integrated into the maintenance track as
> Change \#27391\, perl 5\.8\.9\-to\-be can take the reversion of Change \#20462\.

Thanks for those explanations\. I reverted 20462 in bleadperl as change 28370\.

**Comment by p5pRT** [comment]
### From @ysth

On Sun\, Jun 04\, 2006 at 11&#8203;:26&#8203;:57PM \+0900\, SADAHIRO Tomoyuki wrote&#8203;:
> 
> On Thu\, 01 Jun 2006 11&#8203;:28&#8203;:00 \+0100\, Martin Waite \<martin@&#8203;datacash\.com>
> wrote
> > I wanted to ask why the defined\-ness fragility is only really visible to
> > the XS module I am using \(DBD&#8203;::MySQL\)\. For example Perl's defined
> > operator doesn't have the same problem\.  I looked in the definition of
> > pp\_defined and in this case the Perl defined operator uses the sequence&#8203;:
> > 
> >         if \(SvGMAGICAL\(sv\)\)
> >             mg\_get\(sv\);
> >         if \(SvOK\(sv\)\)
> >             RETPUSHYES;
> > 
> > However\, in the DBD&#8203;::MySQL module\, only the SvOK\(\) test is applied\.
> > 
> > Should the DBD&#8203;::MySQL module be modified to include the MAGIC handling
> > to be more robust ?
> 
> Certainly both perlapi\.pod and perlguts\.pod mentions SvOK\(\) is used
> to check whether an SV is defined\. But it may be necessary to call
> SvGETMAGIC\(\) \(the public macro for \{if \(SvGMAGICAL\(sv\)\) mg\_get\(sv\);\}\)
> before SvOK\(\)\.
> For example\, mg\_get\(\) for a tied scalar corresponds to FETCH\(\) and
> tie\(\) doesn't call mg\_get\(\)\. Hence a tied scalar just after tie\(\)
> isn't true for SvOK\(\)\. But if FETCH\(\) has some side effect\,
> calling too many mg\_get\(\) may cause some unexpected result\. Thus
> I think SvGETMAGIC\(\) should not be used just before each appearance
> of SvOK\(\) or other macros\.

Right\.  XS functions\, just like perl ops\, that don't do just the normal
SvIV/SvPV/etc\. calls \(which the default typemaps usually do\)
should call SvGETMAGIC\(\) once for each input argument before checking
any flags\.

> In my opinion\, the perl's internal is too complicated for XS authors
> to hack and investigate properly\. It may be a workaround to pass simple
> values to XS functions always\. For example\, store the substring into
> a temporary scalar variable and pass the scalar to XS\, rather than pass
> substr\(\) directly to XS\.

XS authors not using the regular typemaps need to know what they are
doing\, but by and large don't\.
